Make the 'config' subcommand a group command

XavierBeguin opened the pull request:
Instead of using options to the 'config' command, use subcommands to
provide more natural ways to request the different subfunctions of this
command.

 - now only runs the gui configuration tool if the 'gui' subcommand is
   provided

 - the '--set' option becomes the 'set' subcommand

 - a new 'show' subcommand is added that shows the value of the given
   configuration parameter(s), or all parameters if no parameter name
   is provided as argument.

ThomasLecocq commented:
Hi Xavier,

Indeed, this is a good idea, making a group out of ``config``. Why ``show`` and not ``get`` ?

Tom

XavierBeguin commented:
Hi Thomas,

I first went with 'get' but it finally appeared it would be more natural to merge it with a 'show' command that would show all parameter values if no parameter name is provided. Simply specifying names of a parameter to 'show' filters the output. It also allows for more code factorisation.

I thought of adding a 'get' subcommand that would invoke the 'show' command, but in my opinion keeping things simple and minimalist is almost always a better option (less user confusion and less code to maintain). YMMV.

Xavier.

ThomasLecocq commented:
ok, I see, and I agree ! (note ``msnoise info`` does already list the parameters, how does this differ?)

XavierBeguin commented:
The main difference is that I did not pay enough attention to `msnoise info` ;) Sorry for not having thought this thoroughly.

I think having `config` being able to display all parameters is nice though, as it's specialised in the handling of the configuration. But that's repetition of functionnality :/  

I like the way `info` shows if a parameter value was changed from its default value. I guess we could either remove one of these lists, or factorise the implementation of this list of parameter values and keep the list of parameters in the output of `info` for completeness (or simply point to the `msnoise config show` option in the `info` output).

Tell me what you prefer, I'll make a new commit.

ThomasLecocq commented:
hehe, thanks for the honnesty.

I'd say it would be best to group things under ``msnoise config`` or ``msnoise db``, with a preference for ``msnoise config``, as the ``db`` group is more intended for import/export and db access...

XavierBeguin commented:
Thanks for you input, Thomas. I wouldn't group it with `db` as this subcommand is indeed more about database management.

After giving it some thoughts, I'm personally not completely sure the output of `info` would fit in `config` either, because of the list of jobs that it shows that is not directly derived from configuration but also from the data scan. However that's a minor issue, and putting it under `msnoise config info` wouldn't shock me too hard either, personally.

In the meantime, I've introduced some changes in commit 8ea3850 to avoid the duplicated functionality introduced by my `msnoise config show`. I've changed it to `msnoise config get <name>` and it now uses a function shared with `info` to display the parameters values (I think that was the best solution). In the process, I'm proposing slight improvements in the code and in the resulting output (the most significant being the printing of the content of the `db.ini`, which is something I find useful). See the commit log message for more details. Please also [see this page](https://framabin.org/p/?79b9798cc27a852f#NJjRR9hxdyQx9o071kH6bzhZx3Fzxkp2l0TxfoefScY=) for an example of output of the modified `msnoise info`.

What do you want to do now ?
 - keep `msnoise info` as an independant subcommand
 - merge it to `config` and create a `msnoise config info` subcommand.

I don't have enough perspective with MSNoise to have a well-founded opinion on this issue. I like and dislike both solutions for different reasons.  The safe bet would be to keep things as they are, I guess.

ThomasLecocq commented:
Xavier,

This looks really good, I like the ``msnoise config get <name>`` like this, and the output of your modified ``msnoise info`` , which in my opinion could remain as is and not be put under msnoise config. 

For me, this is good to merge.
Thomas

XavierBeguin commented:
Cool! Let's keep it like this, then.

As a side note, I forgot to mention that the changes I made to the station information output resolve a tiny bug that was crashing `msnoise info` when a station had no coordinates set (the X, Y, etc. fields were None):

```python
msnoise info
(...)
Stations:
Traceback (most recent call last):
  File "/home/xavier/.virtualenvs/msnoise-dev/bin/msnoise", line 11, in <module>
    load_entry_point('msnoise', 'console_scripts', 'msnoise')()
  File "/home/xavier/Documents/Infra/Msnoise/devel/msnoise.gitlab/msnoise/scripts/msnoise.py", line 961, in run
    cli(obj={})
  File "/home/xavier/.virtualenvs/msnoise-dev/lib/python3.5/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/home/xavier/.virtualenvs/msnoise-dev/lib/python3.5/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/home/xavier/.virtualenvs/msnoise-dev/lib/python3.5/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/xavier/.virtualenvs/msnoise-dev/lib/python3.5/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/xavier/.virtualenvs/msnoise-dev/lib/python3.5/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/home/xavier/Documents/Infra/Msnoise/devel/msnoise.gitlab/msnoise/scripts/msnoise.py", line 269, in info
    print('%s.%s %.4f %.4f %.1f %s %s' % data)
TypeError: a float is required
```

The 'None' value case is now taken into account.

ThomasLecocq commented:
👍 beauuuutiful
